specs/_features/eip7594/das-core.md

```diff
@@ -0,0 +1,222 @@
+
+*Note*: increasing the `custody_size` parameter for a given `node_id` extends the returned list (rather than being an entirely new shuffle) such that if `custody_size` is unknown, the default `CUSTODY_REQUIREMENT` will be correct for a subset of the node's custody.
+
+## Peer discovery
+
+At each slot, a node needs to be able to readily sample from *any* set of columns. To this end, a node should find and maintain a set of diverse and reliable peers that can regularly satisfy their sampling demands.
+
+A node runs a background peer discovery process, maintaining at least `TARGET_NUMBER_OF_PEERS` of various custody distributions (both custody_size and column assignments). The combination of advertised `custody_size` size and public node-id make this readily and publicly accessible.
+
+`TARGET_NUMBER_OF_PEERS` should be tuned upward in the event of failed sampling.
+
+*Note*: while high-capacity and super-full nodes are high value with respect to satisfying sampling requirements, a node should maintain a distribution across node capacities as to not centralize the p2p graph too much (in the extreme becomes hub/spoke) and to distribute sampling load better across all nodes.
+
+*Note*: A DHT-based peer discovery mechanism is expected to be utilized in the above. The beacon-chain network currently utilizes discv5 in a similar method as described for finding peers of particular distributions of attestation subnets. Additional peer discovery methods are valuable to integrate (e.g., latent peer discovery via libp2p gossipsub) to add a defense in breadth against one of the discovery methods being attacked.
+
+## Extended data
+
+In this construction, we extend the blobs using a one-dimensional erasure coding extension. The matrix comprises maximum `MAX_BLOBS_PER_BLOCK` rows and fixed `NUMBER_OF_COLUMNS` columns, with each row containing a `Blob` and its corresponding extension.
+
+## Column gossip
+
+### Parameters
+
+For each column -- use `data_column_sidecar_{subnet_id}` subnets, where each column index maps to the `subnet_id`. The sidecars can be computed with `get_data_column_sidecars(signed_block: SignedBeaconBlock, blobs: Sequence[Blob])` helper.
+
+To custody a particular column, a node joins the respective gossip subnet. Verifiable samples from their respective column are gossiped on the assigned subnet.
+
+### Reconstruction and cross-seeding
+
+If the node obtains 50%+ of all the columns, they can reconstruct the full data matrix via `recover_samples_impl` helper.
+
+If a node fails to sample a peer or fails to get a column on the column subnet, a node can utilize the Req/Resp message to query the missing column from other peers.
+
+Once the node obtain the column, the node should send the missing columns to the column subnets.
```

Comment:
> the node should send the missing columns to the column subnets.

Can we clarify what "should send" means here? I see a few interpretations:
- The node is assigned to a column subnet, but nothing shows up. In this case, the node reconstructs the column and gossip back to the peers on the same column subnet the column blob sidecar
- The node is having trouble with everything because it's only gotten 49% of the columns, it manually requests the last column that made it to 50%, and the node gossip back that last missing 1% column back to the subnet. Not original assigned

Reply:
I think, if node doesn't get a particular `DataColumn` that it needs to sample, it should then attempt to reconstruct the blob using req/resp and then seed it to the network.

with regard to second point, a node will ideally not be requiring 50% columns download unless it chooses to because then there are no actual sampling (if i understand you correctly)

Reply:
> The node is having trouble with everything because it's only gotten 49% of the columns, it manually requests the last column that made it to 50%, and the node gossip back that last missing 1% column back to the subnet. Not original assigned

The node is not supposed to download that many columns as @g11tech said, unless it wants to reconstruct.

> The node is assigned to a column subnet, but nothing shows up. In this case, the node reconstructs the column and gossip back to the peers on the same column subnet the column blob sidecar

Afaiu, we don't have a criteria yet when the node should consider the column as missing.

> I think, if node doesn't get a particular `DataColumn` that it needs to sample, it should then attempt to reconstruct the blob using req/resp and then seed it to the network (that could be the `robustness` bit of the peerDAS)

We also haven't decided yet who will do the reconstruction, since the reconstruction is quite an expensive job.

Reply:
I think for e.g. data availability needs to be verified for the parent block before current block is imported, so the node needs to draw the conclusion before the block import else it can't do the import.

imo the nodes doing the `attestation`/proposal/aggregation would definitely need to do reconstruction else they won't be able to perform their dutites regarding a head that is not available in their forkchoice, but ideally any node could/should do reconstruction and then seed data in their subnets.

so essentially all validator connected nodes should definitely do it as a preparation step for next slot

Reply:
@g11tech would you mind explaining why nodes need to reconstruct the full matrix to perform validator duties?

Reply:
In the current spec, a normal node with `custody_size = 1` cannot help with reconstruction. If a block is > 50% available but not fully available, the assumption is that there is at least one node downloading everything (with `custody_size = 32`, participating in every subnet) which will do reconstruction and re-seed the missing columns *before the next slot*, so that everyone's sampling succeeds and they can import the next proposed block and vote for it

Reply:
> @g11tech would you mind explaining why nodes need to reconstruct the full matrix to perform validator duties?

If sampling fails then availability fails, how can then one import the head block in the forkchoice? so there are two mitigations:

1. Reconstruct by req/rep with root/index and import if one can reconstruct (and also seed the failed sampled but constructed column to its subnet)
2. Or not import the block at all to the forkchoice

Reply:
What does eventually succeed mean? Having retrieved 8 valid samples successfully and _consecutively_ before the DA filter cutoff time?

If it's not consecutive successes, we could eventually get 8 samples even if there are only 8 (out of 128) of them available right?

Reply:
*consecutive* successes is probably too tight and not necessary (given we retry on _the same column_) , as this also relies on the network conditions and the peers.

I think my confusion came from "not retrying on the same column" (randomly pick another column), which doesn't make sense at all 😅 

So I think this is what it means :
- A node has randomly selected to sample 8 columns
- It would try getting samples from those 8 columns
- if it fails to get 1 column from a peer, it would try to retrieve the _same missing column_ with another peer until it finds one
- if the node successfully retrieve samles for the randomly selected 8 columns, DA check is successful, otherwise it fails the DA check.

Please correct me if I'm wrong 🙏

Reply:
Yes, exactly! It's important that you don't just change your choice of columns to sample because of failing. Either you get all of the columns for the random indices you originally chose, or sampling fails

Reply:
Just had a chat with @AgeManning and think we might need some tolerance here with sampling as it's likely that we won't have time to "find peers" if there aren't any peers in a randomly selected column. 

We should have a background task that tries to maintain a uniform set across all columns, but there would still times where we do not have any peers in a given column, and it might be too late to look for a peer as discovery is slow, and even once we find a peer, there's no guarantee that we can connect to them (could be dead, reached peer limit etc). 

Ideally we have some tolerance on sampling here, i.e. rather than requiring 16 out of 16 selected columns to pass, could we require 12 out of 16? (just some make up numbers, could be a higher percentage, but perhaps not 100%)

Reply:
@jimmygchen  @fradamt, to follow up on this, now with the PeerDAS specific numbers:

if you set a false positive threshold of 1e-3, you would need 10 columns if losses are not permitted. Or, you can go for 13 columns, with up to 1 loss. 16 columns with up to 2.
Using incrementalDAS, you can do this as follows:
- you start by doing a random permutation P of the columnIDs. P has 128 elements.
- you request the first 10 columns of the random permutation P[0..9] with a given timeout, say 3 seconds.
- if by 3 seconds all are there, test passed, at minimal network cost.
- if 1 missing, request the missing column from another peer, AND request P[10-12] from peers. All with a new, shorter timeout. Hope to get enough (12 out if 13) by the new timeout.
- and so on ....

Obviously, we can easily adapt this to a lower (more stringent) FP threshold to start with 16 columns.

LossyDAS for PeerDAS calculations: https://colab.research.google.com/drive/18uUgT2i-m3CbzQ5TyP9XFKqTn1DImUJD